src/bgw/job.c

```diff
@@ -1325,9 +1349,11 @@ ts_bgw_job_run_and_set_next_start(BgwJob *job, job_main_func func, int64 initial
 
 		ts_bgw_job_stat_set_next_start(job->fd.id, next_start);
 	}
-	CommitTransactionCommand();
 
-	return ret;
+	if (atomic)
+		CommitTransactionCommand();
```

Comment:
I'm a bit confused: do we have the `next_start` changes to the job coupled with the changes in the job itself?

I think this might also mean that if whatever the function have done can't be committed - then the next_start will also not be updated - meaning that the failing job will be left at most likely at the head of the queue ; probably inhibiting future execution ; I think it may occupy an executor

I think that the transaction should only be opened / closed around the function call.

I think `CommitTransactionCommand()` (or even the FN we are calling) may emit an `ERROR` and I think that should be handled...

I feel like this time changing thing is out-of-place here - I wonder why didn't we just let `ts_bgw_job_stat_mark_end ... get_next_scheduled_execution_slot  ` do its job?

Reply:
> I'm a bit confused: do we have the `next_start` changes to the job coupled with the changes in the job itself?
> 
> I think this might also mean that if whatever the function have done can't be committed - then the next_start will also not be updated - meaning that the failing job will be left at most likely at the head of the queue ; probably inhibiting future execution ; I think it may occupy an executor

As you observed, this is really only used for the telemetry job, so that will just be tried again at a later point.

> I think that the transaction should only be opened / closed around the function call.
> 
> I think `CommitTransactionCommand()` (or even the FN we are calling) may emit an `ERROR` and I think that should be handled...

Hmm... good point.

> I feel like this time changing thing is out-of-place here - I wonder why didn't we just let `ts_bgw_job_stat_mark_end ... get_next_scheduled_execution_slot ` do its job?

Because it causes failures in the tests and I wanted to avoid mixing extensive refactorings with functional changes.

Reply:
> > I think that the transaction should only be opened / closed around the function call.
> > I think `CommitTransactionCommand()` (or even the FN we are calling) may emit an `ERROR` and I think that should be handled...
> 
> Hmm... good point.

Looked into this and it seems there is nothing to do. If an error is thrown, which will be captured by the caller and handled.

Reply:
I see; given that:
* this is only used for telemetry
* telemetry doesn't seem to be doing database changes

I don't think it will matter - we could improve on it later